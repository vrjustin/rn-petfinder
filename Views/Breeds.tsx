import React, {useEffect, useState} from 'react';
import {FlatList, Text, View, StyleSheet, TouchableOpacity} from 'react-native';
import {RouteProp} from '@react-navigation/native';
import {useNavigation} from '@react-navigation/native';
import Breed from '../models/Breed';
import apiService from '../services/apiService';
import FontAwesomeIcon from 'react-native-vector-icons/FontAwesome';

type RootStackParamList = {
  PetTypes: undefined;
  Breeds: {petTypeName: string};
};

type BreedsScreenRouteProp = RouteProp<RootStackParamList, 'Breeds'>;

type Props = {
  route: BreedsScreenRouteProp;
};

const Breeds: React.FC<Props> = ({route}) => {
  const [typeBreeds, setTypeBreeds] = useState<Breed[]>([]);
  const {petTypeName} = route.params;
  const typeName = petTypeName.toLowerCase();
  const navigation = useNavigation();

  useEffect(() => {
    const fetchTypeBreedsData = async () => {
      try {
        const breedsData = await apiService.getPetBreeds(typeName);
        setTypeBreeds(breedsData);
      } catch (error) {
        console.error('Failed top fetch Breeds data: ', error);
      }
    };
    fetchTypeBreedsData();
  }, []);

  const handleBreedSelection = (breed: Breed) => {
    navigation.navigate('Animals', {
      petType: petTypeName,
      selectedBreed: breed,
    });
  };

  const renderItem = ({item}: {item: Breed}) => (
    <TouchableOpacity onPress={() => handleBreedSelection(item)}>
      <View style={styles.item}>
        <FontAwesomeIcon
          name="paw"
          size={20}
          color="#000"
          style={styles.icon}
        />
        <Text style={styles.text}>{item.name}</Text>
      </View>
    </TouchableOpacity>
  );

  return (
    <View style={styles.container}>
      <Text style={styles.header}>
        Select your {typeName} breed, to learn more!
      </Text>
      <FlatList
        data={typeBreeds}
        renderItem={renderItem}
        keyExtractor={item => item.name}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    padding: 20,
    backgroundColor: '#f0f0f0',
  },
  header: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: 20,
  },
  item: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#fff',
    padding: 15,
    marginBottom: 10,
    borderRadius: 5,
    elevation: 3,
  },
  text: {
    fontSize: 16,
  },
  icon: {
    marginRight: 10,
  },
});

export default Breeds;
