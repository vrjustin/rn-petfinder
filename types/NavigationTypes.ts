import {StackNavigationProp} from '@react-navigation/stack';
import {RouteProp, useNavigation} from '@react-navigation/native';
import {PetType} from '../models/PetType';
import Animal from '../models/Animal';
import Breed from '../models/Breed';

type RootStackParamList = {
  PetTypes: undefined;
  Breeds: {petType: PetType};
  Animals: {
    petType: PetType;
    selectedBreeds: Breed[];
    initialIsLoading?: boolean;
  };
  Animal: {selectedAnimal: Animal};
  Options: {from: string};
};

export type BreedsScreenRouteProp = RouteProp<RootStackParamList, 'Breeds'>;
export type AnimalsScreenRouteProp = RouteProp<RootStackParamList, 'Animals'>;
export type AnimalScreenRouteProp = RouteProp<RootStackParamList, 'Animal'>;
export type OptionsScreenRouteProp = RouteProp<RootStackParamList, 'Options'>;

export type BreedsProps = {
  route: BreedsScreenRouteProp;
};

export type AnimalsProps = {
  route: AnimalsScreenRouteProp;
};

export type AnimalProps = {
  route: AnimalScreenRouteProp;
};

export type OptionsProps = {
  route: OptionsScreenRouteProp;
};

export type SignInUpProps = {
  initialLoadingProp?: boolean;
};

export type OrganizationsProps = {
  initialLoadingProp?: boolean;
};

export const useTypedNavigation = () =>
  useNavigation<StackNavigationProp<RootStackParamList>>();
