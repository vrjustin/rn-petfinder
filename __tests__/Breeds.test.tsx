import React from 'react';
import renderer, {act} from 'react-test-renderer';
import Breeds from '../Views/Breeds';
import {BreedsScreenRouteProp} from '../types/NavigationTypes';

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

jest.mock('../services/apiService', () => ({
  getPetBreeds: jest.fn(() => [
    {
      name: 'Bulldog',
      _links: {type: {href: ''}},
    },
    {
      name: 'Labrador',
      _links: {type: {href: ''}},
    },
    {
      name: 'Poodle',
      _links: {type: {href: ''}},
    },
  ]),
}));

describe('Breeds', () => {
  it('renders correctly', async () => {
    let renderedBreedsTree: any;
    await act(async () => {
      renderedBreedsTree = renderer.create(
        <Breeds
          route={{params: {petTypeName: 'Dog'}} as BreedsScreenRouteProp}
        />,
      );
    });
    expect(renderedBreedsTree.toJSON()).toMatchSnapshot();
  });
});
