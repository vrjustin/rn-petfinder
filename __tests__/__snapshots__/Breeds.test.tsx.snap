// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Breeds renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#f0f0f0",
      "flex": 1,
      "padding": 20,
    }
  }
>
  <Text
    style={
      {
        "fontSize": 20,
        "fontWeight": "bold",
        "marginBottom": 20,
      }
    }
  >
    Select your 
    dog
     breed, or start typing below to learn more!
  </Text>
  <TextInput
    onChangeText={[Function]}
    placeholder="Start typing to filter breeds..."
    style={
      {
        "borderColor": "#ddd",
        "borderRadius": 5,
        "borderWidth": 1,
        "height": 40,
        "marginBottom": 10,
        "padding": 10,
      }
    }
    value=""
  />
  <RCTScrollView
    data={
      [
        {
          "_links": {
            "type": {
              "href": "",
            },
          },
          "name": "Bulldog",
        },
        {
          "_links": {
            "type": {
              "href": "",
            },
          },
          "name": "Labrador",
        },
        {
          "_links": {
            "type": {
              "href": "",
            },
          },
          "name": "Poodle",
        },
      ]
    }
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={0.0001}
    stickyHeaderIndices={[]}
    viewabilityConfigCallbackPairs={[]}
  >
    <View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderRadius": 5,
                "elevation": 3,
                "flexDirection": "row",
                "marginBottom": 10,
                "padding": 15,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#000",
                    "fontSize": 20,
                  },
                  {
                    "marginRight": 10,
                  },
                  {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
            <Text
              style={
                {
                  "fontSize": 16,
                }
              }
            >
              Bulldog
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderRadius": 5,
                "elevation": 3,
                "flexDirection": "row",
                "marginBottom": 10,
                "padding": 15,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#000",
                    "fontSize": 20,
                  },
                  {
                    "marginRight": 10,
                  },
                  {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
            <Text
              style={
                {
                  "fontSize": 16,
                }
              }
            >
              Labrador
            </Text>
          </View>
        </View>
      </View>
      <View
        onFocusCapture={[Function]}
        onLayout={[Function]}
        style={null}
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#fff",
                "borderRadius": 5,
                "elevation": 3,
                "flexDirection": "row",
                "marginBottom": 10,
                "padding": 15,
              }
            }
          >
            <Text
              allowFontScaling={false}
              selectable={false}
              style={
                [
                  {
                    "color": "#000",
                    "fontSize": 20,
                  },
                  {
                    "marginRight": 10,
                  },
                  {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  {},
                ]
              }
            >
              
            </Text>
            <Text
              style={
                {
                  "fontSize": 16,
                }
              }
            >
              Poodle
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
